### PR TITLE
sampling_params: add thinking_budget

### PR DESCRIPTION
**Pull request by thyecust**
## Motivation

The deep thinking model sometimes outputs a lengthy reasoning process, which takes a long time to generate the response content and consumes more tokens. To solve this problem, you can set the 
 `thinking_budget` parameter to constrain the maximum length of the reasoning process.

The `thinking_budget` setting does not affect the thinking process of large models. If the number of tokens generated by the model's thinking process exceeds thinking_budget, the reasoning content will be truncated and the final response content will be generated immediately.

## Modifications

1. Add `thinking_budget` parameter. Can be used in `extra_body={}` (like [AlibabaCloud doc](alibabacloud.com/help/zh/model-studio/deep-thinking#:~:text="thinking_budget"%3A%2050)) or `sampling_params` (see `docs/backend/sampling_params.md`)
2. Add new fields for `SamplingBatchInfo`:
  - `thinking_budgets: torch.Tensor`: thinking_budget for r in reqs, -1 if no budget.
  - `num_thinking_tokens: torch.Tensor`: length of thinking process for r in reqs.
  - `think_end_ids: torch.Tensor`: encode(think_end) for r in reqs.
3. In `ModelRunner.sample()`
  - before preprocessing logits, increase `num_thinking_tokens`, mask all tokens except think_end if `num_thinking_tokens > thinking_budgets`
  - after sampled `next_token`, set `thinking_budget` to `-1` if thinking ends.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by thyecust** [comment]
Tested by

```python
from openai import OpenAI

openai_api_key = "EMPTY"
openai_api_base = f"http://127.0.0.1:30000/v1"
client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

model = "Qwen/Qwen3-8B"  # Use the model loaded by the server
messages = [{"role": "user", "content": "9.11 and 9.8, which is greater?"}]

response = client.chat.completions.create(
    model=model,
    messages=messages,
    extra_body={
        "chat_template_kwargs": {"enable_thinking": True},
        "separate_reasoning": True,
        "thinking_budget": 100,
    },
)

print(
    "response.choices[0].message.reasoning_content: \n",
    response.choices[0].message.reasoning_content,
)
print("response.choices[0].message.content: \n", response.choices[0].message.content)
```

See also: https://www.alibabacloud.com/help/zh/model-studio/deep-thinking#:~:text=%22thinking_budget%22%3A%2050

**Comment by lambert0312** [comment]
Although the think part is shortened, the original reasoning content will continue to be output, resulting in two </think> as a whole. It seems that the current version does not really truncate the output of the thinking content.

**Comment by thyecust** [comment]
> Although the think part is shortened, the original reasoning content will continue to be output, resulting in two </think> as a whole. It seems that the current version does not really truncate the output of the thinking content.

I think this is the characteristic of model itself (how to model a language after `</think>` token). Not a problem for inference.

**Comment by thyecust** [comment]
@minleminzui the PR is ready for review. Two falied tests are minor.

**Comment by minleminzui** [comment]
> @minleminzui the PR is ready for review. Two falied tests are minor.

LGTM, I checked your code, it works

**Review by minleminzui — APPROVED** [review]
